### PR TITLE
Add a new blog 0323

### PR DESCRIPTION
**Pull request by booooodv**

### CODE REVIEW

**Comment by booooodv** [comment]
@husiyu PTAL

**Comment by husiyu** [comment]
LGTM

File: tikv-source-code-reading-18.md [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+summary: raft-rs 实现了 Raft Leader election 和 Log replication 等核心功能，而消息的发送、接收、应用到状态机等操作则需要使用者自行实现，本文将要介绍的就是 TiKV 中这些部分的处理过程。
+tags: ['TiKV 源码解析','社区']
+---
+在学习了 [前面的文章](https://pingcap.com/blog-cn/tikv-source-code-reading-2/) 之后，相信大家已经对 TiKV 使用的 Raft 核心库 **raft-rs** 有了基本的了解。**raft-rs** 实现了 Raft Leader election 和 Log replication 等核心功能，而消息的发送、接收、应用到状态机等操作则需要使用者自行实现，本文将要介绍的就是 TiKV 中这些部分的处理过程。
+
+## Raft Ready
+
+在开始正题之前，我们先简单回顾一下 **raft-rs** 与外部代码的交互接口: Ready。 Ready 结构的 [定义](https://github.com/tikv/raft-rs/blob/afabefa23196a4a23832add0087f5a522d8ccc3b/src/raw_node.rs#L86-L110) 如下：
+
+```
```
Comment (5kbpers) [inline tikv-source-code-reading-18.md]:
```suggestion
```rust
```

File: tikv-source-code-reading-18.md [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+
+* **RaftMessage**: 其他 Peer 发送过来 Raft 消息，包括心跳、日志、投票消息等。
+
+* **RaftCommand**: 上层提出的 proposal，其中包含了需要通过 Raft 同步的操作，以及操作成功之后需要调用的 `callback` 函数。
+
+* **ApplyRes**: ApplyFsm 在将日志应用到状态机之后发送给 `PeerFsm` 的消息，用于在进行操作之后更新某些内存状态。
+
+我们主要关注的是 `PeerFsm` 如何处理 Proposal，也就是 `RaftCommand` 的处理过程。在进入到 `PeerFsmDelegate::propose_raft_command` 后，首先会调用 `PeerFsmDelegate::pre_propose_raft_command` 对 peer ID, peer term, region epoch (region 的版本，region split、merge 和 add / delete peer 等操作会改变 region epoch) 是否匹配、 peer 是否 leader 等条件进行一系列检查，并根据请求的类型（是读请求还是写请求），选择不同的 Propose 策略见（ `Peer::inspect`）：
+
+```
```
Comment (5kbpers) [inline tikv-source-code-reading-18.md]:
```suggestion
```rust
```

File: tikv-source-code-reading-18.md [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+
+## Proposal 的接收和在 Raft 中的复制
+
+TiKV 3.0 中引入了类似 [Actor](https://en.wikipedia.org/wiki/Actor_model) 的并发模型，Actor 被视为并发运算的基本单元：当一个 Actor 接收到一则消息，它可以做出一些决策、创建更多的 Actor、发送更多的消息、决定要如何回答接下来的消息。每个 TiKV 上的 Raft Peer 都对应两个 Actor，我们把它们分别称为 `PeerFsm` 和 `ApplyFsm`。`PeerFsm` 用于接收和处理其他 Raft Peer 发送过来的 Raft 消息，而 `ApplyFsm` 用于将已提交日志应用到状态机。
+
+TiKV 中实现的 Actor System 被称为 BatchSystem，它使用几个 Poll 线程从多个 Mailbox 上拉取一个 Batch 的消息，再分别交由各个 Actor 来执行。为了保证 [线性一致性](https://pingcap.com/blog-cn/linearizability-and-raft/) ，一个 Actor 同时只会在一个 Poll 线程上接收消息并顺序执行。由于篇幅所限，这一部分的实现在这里不做详述，感兴趣的同学可以在 `raftstore/fsm/batch.rs` 查看详细代码。
+
+上面谈到，`PeerFsm` 用于接收和处理 Raft 消息。它接收的消息为 `PeerMsg`，根据消息类型的不同会有不同的处理：
+
+```
```
Comment (5kbpers) [inline tikv-source-code-reading-18.md]:
```suggestion
```rust
```

File: tikv-source-code-reading-18.md (hunk 1 above)
Comment (booooodv) [reply tikv-source-code-reading-18.md]: done

File: tikv-source-code-reading-18.md (hunk 2 above)
Comment (booooodv) [reply tikv-source-code-reading-18.md]: done

File: tikv-source-code-reading-18.md (hunk 3 above)
Comment (booooodv) [reply tikv-source-code-reading-18.md]: done

**Pull request merged**